### DISCUSSION TITLE
Is there any way to extract lora adapter from merged model?

### DISCUSSION THREAD

**Discussion by tjtanaa** [huggingface/peft#1219] [Q&A]
Is there any way to extract lora adapter from merged model?

Recently, there have been a marvelous improvement in deploying multiple LoRA. I wonder is there any existing code that could be use to extract the merged weights back into LoRA adapter to make use of the existing multi lora inference engine, e.g. lorax and storage.

**Comment by younesbelkada** [comment]
Technically I think this is possible, you can merge adapters that you want to merge by specifying `adapter_names` in `merge`: https://github.com/huggingface/peft/blob/bffbbbf76ab3af4563fa4c6079840d19dc423a9b/src/peft/tuners/lora/layer.py#L267 - after that you can have access to the list of merged adapters inside the `merged_adapters` attribute of `LoraLayer`: https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/layer.py#L48 

You can simply use that list to retrieve LoRA weights from `lora_A` and `lora_B` module dicts as their weights will still remain there after merging (and not unloading). If you merge + unload the model it might be complex to retrieve the merged lora weights

**Reply by BenjaminBossan** [reply]
To add to Younes' response.

> to extract the merged weights back into LoRA adapter to make use of the existing multi lora inference engine

If your use case is that you have multiple LoRA adapters and you want to merge adapter 1 for faster inference, but later switch to adapter 2 and merge that, later to adapter 3 etc. here is how I would do it.

1. Load all adapters. At this point, adapter 1 (the first one loaded) should be the `active_adapter`.
2. Call  `merge_adapter`, which should merge the active adapter, in this case 1 (don't use `merge_and_unload`, as Younes mentioned)
3. Once you're finished with that adapter, call `unmerge_adapter`
4. Switch to adapter 2 by calling `set_adapter`; go back to step 2.

And for future reference, the permalink of Younes' link: https://github.com/huggingface/peft/blob/c22a8e5d47774b3c9ed7352994ccd34e757fff58/src/peft/tuners/lora/layer.py#L48

**Comment by tjtanaa** [comment]
> If you merge + unload the model it might be complex to retrieve the merged lora weights

What I am currently looking for is this option as there are many models e.g. WizardLM, dolphine, openhermes, etc are all models with merged lora weights (merge + unload). I am interesting in this as s-lora and lora-x helps to schedule how the lora adapters will be used, and I only need to host one base model and multiple lora (which takes up significantly smaller amount of GPU memory).

**Reply by BenjaminBossan** [reply]
As Younes mentioned, if the adapters are completely merged into the base model weights, it's not trivial to extract them again. There is nothing in PEFT that supports this, you would have to do this manually, and it would require access to both the merged weights + adapter weights separately, or the merged weights + base weights separately.

**Comment by tanyav2** [comment]
You could use `get_peft_model_state_dict` to extract the adapter weights. It's also discussed in [PEFT low level API](https://huggingface.co/docs/peft/developer_guides/low_level_api).

https://github.com/huggingface/peft/blob/26504a01199e1204945bfcd9b3f29458d99b4e59/src/peft/utils/save_and_load.py#L41-L43

**Comment by thomasgauthier** [comment]
I implemented something in this direction using singular value decomposition (SVD). I call it [LoRD](https://github.com/thomasgauthier/LoRD) for *Lo*w-*R*ank *D*ecomposition
